File: android/src/main/java/com/indigoviolet/posenet/PosenetDecoder.java
package com.indigoviolet.posenet;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.PriorityQueue;

public class PosenetDecoder {

    private String[] partNames = {
        "nose", "leftEye", "rightEye", "leftEar", "rightEar", "leftShoulder",
        "rightShoulder", "leftElbow", "rightElbow", "leftWrist", "rightWrist",
        "leftHip", "rightHip", "leftKnee", "rightKnee", "leftAnkle", "rightAnkle"
    };

    private String[][] poseChain = {
        {"nose", "leftEye"}, {"leftEye", "leftEar"}, {"nose", "rightEye"},
        {"rightEye", "rightEar"}, {"nose", "leftShoulder"},
        {"leftShoulder", "leftElbow"}, {"leftElbow", "leftWrist"},
        {"leftShoulder", "leftHip"}, {"leftHip", "leftKnee"},
        {"leftKnee", "leftAnkle"}, {"nose", "rightShoulder"},
        {"rightShoulder", "rightElbow"}, {"rightElbow", "rightWrist"},
        {"rightShoulder", "rightHip"}, {"rightHip", "rightKnee"},
        {"rightKnee", "rightAnkle"}
    };

    private int mOutputStride;
    private int mNumKeypoints;

    public PosenetDecoder(int outputStride) {
        mOutputStride = outputStride;
        initPoseNet();
    }

    private Map<String, Integer> partsIds = new HashMap<>();
    private List<Integer> parentToChildEdges = new ArrayList<>();
    private List<Integer> childToParentEdges = new ArrayList<>();

    private void initPoseNet() {
        mNumKeypoints = partNames.length;
        for (int i = 0; i < mNumKeypoints; ++i)
            partsIds.put(partNames[i], i);

        for (String[] edge : poseChain) {
            parentToChildEdges.add(partsIds.get(edge[1]));
            childToParentEdges.add(partsIds.get(edge[0]));
        }
    }

    private Map<String, Object> makeKeypoint(float score, int partId, float y, float x) {
        Map<String, Object> keypoint = new HashMap<>();
        keypoint.put("score", score);
        keypoint.put("partId", partId);
        keypoint.put("part", partNames[partId]);

        Map<String, Float> position = new HashMap<>();
        position.put("y", y);
        position.put("x", x);
        keypoint.put("position", position);

        // keypoint.put("y", y);
        // keypoint.put("x", x);
        return keypoint;
    }

    public List<Map<String, Object>> decode(Map<Integer, Object> outputMap, final int numResults, final float threshold, final int nmsRadius) {
        int localMaximumRadius = 1;

        float[][][] scores = ((float[][][][]) outputMap.get(0))[0];
        float[][][] offsets = ((float[][][][]) outputMap.get(1))[0];
        float[][][] displacementsFwd = ((float[][][][]) outputMap.get(2))[0];
        float[][][] displacementsBwd = ((float[][][][]) outputMap.get(3))[0];

        PriorityQueue<Map<String, Object>> pq = buildPartWithScoreQueue(scores, threshold, localMaximumRadius);

        int squaredNmsRadius = nmsRadius * nmsRadius;

        List<Map<String, Object>> poses = new ArrayList<>();
        while (poses.size() < numResults && pq.size() > 0) {
            Map<String, Object> root = pq.poll();
            float[] rootPoint = getImageCoords(root, offsets);

            if (withinNmsRadiusOfCorrespondingPoint(poses, squaredNmsRadius, rootPoint[0], rootPoint[1], (int) root.get("partId")))
                continue;

            List<Map<String, Object>> keypoints = decodePose(root, rootPoint, scores, offsets, displacementsFwd, displacementsBwd);
            float score = getInstanceScore(keypoints, poses, squaredNmsRadius);

            Map<String, Object> pose = new HashMap<>();
            pose.put("keypoints", keypoints);
            pose.put("score", score);
            poses.add(pose);
        }
        return poses;
    }

    private List<Map<String, Object>> decodePose(Map<String, Object> root, float[] rootPoint, float[][][] scores, float[][][] offsets, float[][][] displacementsFwd, float[][][] displacementsBwd) {
        int numEdges = parentToChildEdges.size();
        Map<String, Object> rootKeypoint = makeKeypoint((float) root.get("score"), (int) root.get("partId"), rootPoint[0], rootPoint[1]);

        Map<Integer, Map<String, Object>> keypoints = new HashMap<>();
        keypoints.put((int) root.get("partId"), rootKeypoint);

        for (int edge = numEdges - 1; edge >= 0; --edge) {
            int sourceKeypointId = parentToChildEdges.get(edge);
            int targetKeypointId = childToParentEdges.get(edge);
            if (keypoints.containsKey(sourceKeypointId) && !keypoints.containsKey(targetKeypointId)) {
                Map<String, Object> keypoint = traverseToTargetKeypoint(edge, keypoints.get(sourceKeypointId), targetKeypointId, scores, offsets, displacementsBwd);
                keypoints.put(targetKeypointId, keypoint);
            }
        }

        for (int edge = 0; edge < numEdges; ++edge) {
            int sourceKeypointId = childToParentEdges.get(edge);
            int targetKeypointId = parentToChildEdges.get(edge);
            if (keypoints.containsKey(sourceKeypointId) && !keypoints.containsKey(targetKeypointId)) {
                Map<String, Object> keypoint = traverseToTargetKeypoint(edge, keypoints.get(sourceKeypointId), targetKeypointId, scores, offsets,  displacementsFwd);
                keypoints.put(targetKeypointId, keypoint);
            }
        }
        return new ArrayList<>(keypoints.values());
    }

    private PriorityQueue<Map<String, Object>> buildPartWithScoreQueue(float[][][] scores, double threshold, int localMaximumRadius) {
        PriorityQueue<Map<String, Object>> pq =
            new PriorityQueue<>(
                                1,
                                new Comparator<Map<String, Object>>() {
                                    @Override
                                    public int compare(Map<String, Object> lhs, Map<String, Object> rhs) {
                                        return Float.compare((float) rhs.get("score"), (float) lhs.get("score"));
                                    }
                                });

        for (int heatmapY = 0; heatmapY < scores.length; ++heatmapY) {
            for (int heatmapX = 0; heatmapX < scores[0].length; ++heatmapX) {
                for (int keypointId = 0; keypointId < scores[0][0].length; ++keypointId) {
                    float score = sigmoid(scores[heatmapY][heatmapX][keypointId]);
                    if (score < threshold) continue;

                    if (scoreIsMaximumInLocalWindow(keypointId, score, heatmapY, heatmapX, localMaximumRadius, scores)) {
                        Map<String, Object> res = makeKeypoint(score, keypointId, heatmapY, heatmapX);
                        pq.add(res);
                    }
                }
            }
        }

        return pq;
    }

    private boolean scoreIsMaximumInLocalWindow(int keypointId,
                                                float score,
                                                int heatmapY,
                                                int heatmapX,
                                                int localMaximumRadius,
                                                float[][][] scores) {
        boolean localMaximum = true;
        int height = scores.length;
        int width = scores[0].length;

        int yStart = Math.max(heatmapY - localMaximumRadius, 0);
        int yEnd = Math.min(heatmapY + localMaximumRadius + 1, height);
        for (int yCurrent = yStart; yCurrent < yEnd; ++yCurrent) {
            int xStart = Math.max(heatmapX - localMaximumRadius, 0);
            int xEnd = Math.min(heatmapX + localMaximumRadius + 1, width);
            for (int xCurrent = xStart; xCurrent < xEnd; ++xCurrent) {
                if (sigmoid(scores[yCurrent][xCurrent][keypointId]) > score) {
                    localMaximum = false;
                    break;
                }
            }
            if (!localMaximum) {
                break;
            }
        }

        return localMaximum;
    }

    private float[] getOffsetPoint(int y, int x, int keypointId, float[][][] offsets) {
        float offsetY = offsets[y][x][keypointId];
        float offsetX = offsets[y][x][keypointId + mNumKeypoints];
        return new float[]{offsetY, offsetX};
    }

    private float[] getImageCoords(Map<String, Object> keypoint, float[][][] offsets) {
        // This is only invoked from keypoints that are on the
        // PriorityQueue, where x and y are integers, so it's safe to
        // round them (we can't cast float to int)
        int heatmapY = Math.round(getKeypointPosition(keypoint, "y"));
        int heatmapX = Math.round(getKeypointPosition(keypoint, "x"));
        int keypointId = (int) keypoint.get("partId");
        float[] offsetPoint = getOffsetPoint(heatmapY, heatmapX, keypointId, offsets);
        // int keypointId = (int) keypoint.get("partId");
        // float offsetY = offsets[heatmapY][heatmapX][keypointId];
        // float offsetX = offsets[heatmapY][heatmapX][keypointId + numParts];

        float y = heatmapY * mOutputStride + offsetPoint[0];
        float x = heatmapX * mOutputStride + offsetPoint[1];

        return new float[]{y, x};
    }

    private float squaredDistance(float y1, float x1, float y2, float x2) {
        float dy = y1 - y2;
        float dx = x1 - x2;
        return dy * dy + dx * dx;
    }

    private boolean withinNmsRadiusOfCorrespondingPoint(List<Map<String, Object>> poses,
                                                        float squaredNmsRadius,
                                                        float y,
                                                        float x,
                                                        int keypointId) {
        for (Map<String, Object> pose : poses) {
            List<Map<String, Object>> keypoints = (List<Map<String, Object>>) pose.get("keypoints");
            Map<String, Object> correspondingKeypoint = keypoints.get(keypointId);
            float sq = squaredDistance(y, x, getKeypointPosition(correspondingKeypoint, "y"),  getKeypointPosition(correspondingKeypoint, "x"));
            if (sq <= squaredNmsRadius)
                return true;
        }

        return false;
    }

    private float getKeypointPosition(Map<String, Object> keypoint, String axis) {
        return ((Map<String, Float>) keypoint.get("position")).get(axis);
    }



    private Map<String, Object> traverseToTargetKeypoint(int edgeId,
                                                         Map<String, Object> sourceKeypoint,
                                                         int targetKeypointId,
                                                         float[][][] scores,
                                                         float[][][] offsets,
                                                         float[][][] displacements) {
        int height = scores.length;
        int width = scores[0].length;
        float sourceKeypointY = getKeypointPosition(sourceKeypoint, "y");
        float sourceKeypointX = getKeypointPosition(sourceKeypoint, "x");

        int[] sourceKeypointIndices = getStridedIndexNearPoint(sourceKeypointY, sourceKeypointX, height, width);

        float[] displacement = getDisplacement(edgeId, sourceKeypointIndices, displacements);

        float[] displacedPoint = new float[]{
            sourceKeypointY + displacement[0],
            sourceKeypointX + displacement[1]
        };

        float[] targetKeypoint = displacedPoint;

        final int offsetRefineStep = 2;
        for (int i = 0; i < offsetRefineStep; i++) {
            int[] targetKeypointIndices = getStridedIndexNearPoint(targetKeypoint[0], targetKeypoint[1], height, width);

            int targetKeypointY = targetKeypointIndices[0];
            int targetKeypointX = targetKeypointIndices[1];

            float[] offsetPoint = getOffsetPoint(targetKeypointY, targetKeypointX, targetKeypointId, offsets);

            targetKeypoint = new float[]{
                targetKeypointY * mOutputStride + offsetPoint[0],
                targetKeypointX * mOutputStride + offsetPoint[1]
            };
        }

        int[] targetKeypointIndices = getStridedIndexNearPoint(targetKeypoint[0], targetKeypoint[1], height, width);

        float score = sigmoid(scores[targetKeypointIndices[0]][targetKeypointIndices[1]][targetKeypointId]);

        return makeKeypoint(score, targetKeypointId, targetKeypoint[0], targetKeypoint[1] );
    }

    private int clamp(int v, int min, int max) {
        if (v < min) return min;
        if (v > max) return max;
        return v;
    }

    private int[] getStridedIndexNearPoint(float _y, float _x, int height, int width) {
        int y = clamp(Math.round(_y / mOutputStride), 0, height - 1);
        int x = clamp(Math.round(_x / mOutputStride), 0, width - 1);
        return new int[]{y, x};
    }

    private float[] getDisplacement(int edgeId, int[] keypoint, float[][][] displacements) {
        int numEdges = displacements[0][0].length / 2;
        int y = keypoint[0];
        int x = keypoint[1];
        return new float[]{displacements[y][x][edgeId], displacements[y][x][edgeId + numEdges]};
    }

    private float getInstanceScore(List<Map<String, Object>> keypoints, List<Map<String, Object>> existingPoses, int squaredNmsRadius) {
        float scores = 0;
        for (Map<String, Object> keypoint : keypoints) {
            if (withinNmsRadiusOfCorrespondingPoint(existingPoses, squaredNmsRadius, getKeypointPosition(keypoint, "y"), getKeypointPosition(keypoint, "x"), (int) keypoint.get("partId")))
                continue;
            scores += (float) keypoint.get("score");
        }

        return scores / mNumKeypoints;
    }

    private float sigmoid(final float x) {
        return (float) (1. / (1. + Math.exp(-x)));
    }
}

File: android/src/main/java/com/indigoviolet/posedecoding/PosenetDecoder.kt
package com.indigoviolet.posedecoding

import java.util.*
import kotlin.math.exp
import kotlin.math.roundToInt

typealias ThreeDFloatArray = Array<Array<FloatArray>>
typealias FourDFloatArray = Array<ThreeDFloatArray>

class PosenetDecoder private constructor(
    private val model: TfliteModel,
    private val params: DecoderParams
) : Decoder(model, params) {

    companion object {
        private lateinit var instance: PosenetDecoder

        @JvmStatic
        fun getInstance(model: TfliteModel, params: DecoderParams): PosenetDecoder {
            // This is only to try to keep the same instance shared
            // across multiple invocations of decode, to avoid
            // reinitialization. It's not important that this be
            // synchronized -- there is no actual shared state across
            // threads
            val existsAndMatches = ::instance.isInitialized && (
                instance.model == model &&
                instance.params == params
            )
            if (!existsAndMatches) {
                instance = PosenetDecoder(model, params)
            }

            return instance
        }
    }

    private val partNames = arrayOf("nose",
            "leftEye", "rightEye",
            "leftEar", "rightEar",
            "leftShoulder", "rightShoulder",
            "leftElbow", "rightElbow",
            "leftWrist", "rightWrist",
            "leftHip", "rightHip",
            "leftKnee", "rightKnee",
            "leftAnkle", "rightAnkle")

    private val poseChain = arrayOf(
            Pair("nose", "leftEye"),
            Pair("leftEye", "leftEar"),
            Pair("nose", "rightEye"),
            Pair("rightEye", "rightEar"),
            Pair("nose", "leftShoulder"),
            Pair("leftShoulder", "leftElbow"),
            Pair("leftElbow", "leftWrist"),
            Pair("leftShoulder", "leftHip"),
            Pair("leftHip", "leftKnee"),
            Pair("leftKnee", "leftAnkle"),
            Pair("nose", "rightShoulder"),
            Pair("rightShoulder", "rightElbow"),
            Pair("rightElbow", "rightWrist"),
            Pair("rightShoulder", "rightHip"),
            Pair("rightHip", "rightKnee"),
            Pair("rightKnee", "rightAnkle")
    )

    private val partsIds: Map<String, Int>
    private val parentToChildEdges: List<Int>
    private val childToParentEdges: List<Int>

    init {
        partsIds = partNames.indices.map { Pair(partNames[it], it) }.toMap()
        parentToChildEdges = poseChain.map { partsIds.getValue(it.second) }
        childToParentEdges = poseChain.map { partsIds.getValue(it.first) }
    }

    override fun addPixelValue(v: Int) {
        inputBuf.putFloat((((v shr 16) and 0xFF) - params.posenetMean) / params.posenetStd)
        inputBuf.putFloat((((v shr 8) and 0xFF) - params.posenetMean) / params.posenetStd)
        inputBuf.putFloat(((v and 0xFF) - params.posenetMean) / params.posenetStd)
    }

    override fun decode(): List<Pose> {
        val localMaximumRadius = 1

        @Suppress("UNCHECKED_CAST")
        val scores = (outputBuf[0] as FourDFloatArray)[0]

        @Suppress("UNCHECKED_CAST")
        val offsets = (outputBuf[1] as FourDFloatArray)[0]

        @Suppress("UNCHECKED_CAST")
        val displacementsFwd = (outputBuf[2] as FourDFloatArray)[0]

        @Suppress("UNCHECKED_CAST")
        val displacementsBwd = (outputBuf[3] as FourDFloatArray)[0]

        val pq = buildPartWithScoreQueue(scores, params.threshold.toDouble(), localMaximumRadius)

        val squaredNmsRadius = params.posenetNMSRadius * params.posenetNMSRadius

        val poses = mutableListOf<Pose>()
        while (poses.size < params.posenetNumResults && pq.size > 0) {
            val root = pq.poll()
            val rootPoint = getImageCoords(root, offsets)

            if (withinNmsRadiusOfCorrespondingPoint(poses, squaredNmsRadius.toFloat(), rootPoint, root.partId))
                continue

            val keypoints = decodePose(root, rootPoint, scores, offsets, displacementsFwd, displacementsBwd)
            val score = getInstanceScore(keypoints, poses, squaredNmsRadius)
            poses.add(Pose(score, keypoints))
        }
        return poses
    }

    private fun decodePose(root: Keypoint, rootPoint: FloatCoords, scores: ThreeDFloatArray, offsets: ThreeDFloatArray, displacementsFwd: ThreeDFloatArray, displacementsBwd: ThreeDFloatArray): List<Keypoint> {
        val rootKeypoint = Keypoint(partNames[root.partId], root.partId, rootPoint, root.score)

        val keypoints = mutableMapOf<Int, Keypoint>()
        keypoints[root.partId] = rootKeypoint

        for (edge in parentToChildEdges.indices.reversed()) {
            val sourceKeypointId = parentToChildEdges[edge]
            val targetKeypointId = childToParentEdges[edge]
            if (keypoints.containsKey(sourceKeypointId) && !keypoints.containsKey(targetKeypointId)) {
                val keypoint = traverseToTargetKeypoint(edge, keypoints[sourceKeypointId]!!, targetKeypointId, scores, offsets, displacementsBwd)
                keypoints[targetKeypointId] = keypoint
            }
        }

        for (edge in parentToChildEdges.indices) {
            val sourceKeypointId = childToParentEdges[edge]
            val targetKeypointId = parentToChildEdges[edge]
            if (keypoints.containsKey(sourceKeypointId) && !keypoints.containsKey(targetKeypointId)) {
                val keypoint = traverseToTargetKeypoint(edge, keypoints[sourceKeypointId]!!, targetKeypointId, scores, offsets, displacementsFwd)
                keypoints[targetKeypointId] = keypoint
            }
        }
        return keypoints.values.toList()
    }

    private fun buildPartWithScoreQueue(scores: ThreeDFloatArray, threshold: Double, localMaximumRadius: Int): PriorityQueue<Keypoint> {
        val pq = PriorityQueue(1, compareByDescending<Keypoint> { it.score })

        for (heatmapY in scores.indices) {
            for (heatmapX in scores[0].indices) {
                for (keypointId in scores[0][0].indices) {
                    val score = sigmoid(scores[heatmapY][heatmapX][keypointId])
                    if (score < threshold) continue

                    if (scoreIsMaximumInLocalWindow(keypointId, score, IntCoords(heatmapY, heatmapX), localMaximumRadius, scores)) {
                        val res = Keypoint(partNames[keypointId], keypointId, FloatCoords(heatmapY.toFloat(), heatmapX.toFloat()), score)
                        pq.add(res)
                    }
                }
            }
        }

        return pq
    }

    private fun scoreIsMaximumInLocalWindow(keypointId: Int,
                                            score: Float,
                                            heatmapIdxs: IntCoords,
                                            localMaximumRadius: Int,
                                            scores: ThreeDFloatArray): Boolean {
        var localMaximum = true
        val height = scores.size
        val width = scores[0].size

        val yStart = maxOf(heatmapIdxs.y - localMaximumRadius, 0)
        val yEnd = minOf(heatmapIdxs.y + localMaximumRadius + 1, height)
        for (yCurrent in yStart until yEnd) {
            val xStart = maxOf(heatmapIdxs.x - localMaximumRadius, 0)
            val xEnd = minOf(heatmapIdxs.x + localMaximumRadius + 1, width)
            for (xCurrent in xStart until xEnd) {
                if (sigmoid(scores[yCurrent][xCurrent][keypointId]) > score) {
                    localMaximum = false
                    break
                }
            }
            if (!localMaximum) {
                break
            }
        }

        return localMaximum
    }

    private fun getOffsetPoint(pointIdxs: IntCoords, keypointId: Int, offsets: ThreeDFloatArray): FloatCoords {
        val (y, x) = pointIdxs
        val offsetY = offsets[y][x][keypointId]
        val offsetX = offsets[y][x][keypointId + partNames.size]
        return FloatCoords(offsetY, offsetX)
    }

    private fun getImageCoords(keypoint: Keypoint, offsets: ThreeDFloatArray): FloatCoords {
        // This is only invoked from keypoints that are on the
        // PriorityQueue, where x and y are integers, so it's safe to
        // round them (we can't cast float to int)
        val heatmapY = keypoint.y.roundToInt()
        val heatmapX = keypoint.x.roundToInt()
        val keypointId = keypoint.partId
        val (offsetY, offsetX) = getOffsetPoint(IntCoords(heatmapY, heatmapX), keypointId, offsets)
        val y = heatmapY * model.posenetOutputStride + offsetY
        val x = heatmapX * model.posenetOutputStride + offsetX

        return FloatCoords(y, x)
    }

    private fun squaredDistance(left: FloatCoords, right: FloatCoords): Float {
        val dy = left.y - right.y
        val dx = left.x - right.x
        return dy * dy + dx * dx
    }

    private fun withinNmsRadiusOfCorrespondingPoint(poses: List<Pose>,
                                                    squaredNmsRadius: Float,
                                                    point: FloatCoords,
                                                    keypointId: Int): Boolean {
        for (pose in poses) {
            val keypoints = pose.keypoints
            val sq = squaredDistance(point, keypoints[keypointId].coords)
            if (sq <= squaredNmsRadius)
                return true
        }

        return false
    }

    private fun traverseToTargetKeypoint(edgeId: Int,
                                         sourceKeypoint: Keypoint,
                                         targetKeypointId: Int,
                                         scores: ThreeDFloatArray,
                                         offsets: ThreeDFloatArray,
                                         displacements: ThreeDFloatArray): Keypoint {
        val height = scores.size
        val width = scores[0].size
        val sourceKeypointIndices = getStridedIndexNearPoint(sourceKeypoint.coords, height, width)

        val (displacementY, displacementX) = getDisplacement(edgeId, sourceKeypointIndices, displacements)
        val displacedPoint = FloatCoords(sourceKeypoint.y + displacementY, sourceKeypoint.x + displacementX)
        var targetKeypoint = displacedPoint

        val offsetRefineStep = 2
        for (i in 0 until offsetRefineStep) {
            val targetKPIdxs = getStridedIndexNearPoint(targetKeypoint, height, width)
            val (offsetY, offsetX) = getOffsetPoint(targetKPIdxs, targetKeypointId, offsets)

            targetKeypoint = FloatCoords(targetKPIdxs.y * model.posenetOutputStride + offsetY, targetKPIdxs.x * model.posenetOutputStride + offsetX)
        }

        val (targetKPIdxY, targetKPIdxX) = getStridedIndexNearPoint(targetKeypoint, height, width)
        val score = sigmoid(scores[targetKPIdxY][targetKPIdxX][targetKeypointId])
        return Keypoint(partNames[targetKeypointId], targetKeypointId, targetKeypoint, score)
    }

    private fun clamp(v: Int, min: Int, max: Int): Int {
        return when {
            v < min -> min
            v > max -> max
            else -> v
        }
    }

    private fun getStridedIndexNearPoint(kp: FloatCoords, height: Int, width: Int): IntCoords {
        val (_y, _x) = kp
        val y = clamp((_y / model.posenetOutputStride).roundToInt(), 0, height - 1)
        val x = clamp((_x / model.posenetOutputStride).roundToInt(), 0, width - 1)
        return IntCoords(y, x)
    }

    private fun getDisplacement(edgeId: Int, keypointIdxs: IntCoords, displacements: ThreeDFloatArray): FloatCoords {
        val numEdges = displacements[0][0].size / 2
        val (y, x) = keypointIdxs
        return FloatCoords(displacements[y][x][edgeId], displacements[y][x][edgeId + numEdges])
    }

    private fun getInstanceScore(keypoints: List<Keypoint>, existingPoses: List<Pose>, squaredNmsRadius: Int): Float {
        var scores = 0f
        for ((_, partId, coords, score) in keypoints) {
            if (withinNmsRadiusOfCorrespondingPoint(existingPoses, squaredNmsRadius.toFloat(), coords, partId))
                continue
            scores += score
        }

        return scores / partNames.size
    }

    private fun sigmoid(x: Float): Float {
        return (1.0f / (1.0f + exp(-1.0f * x)))
    }
}
